File: src/main/java/refactown/cleancode/c06objects/envy/FeatureEnvy.java
package refactown.cleancode.c06objects.envy;

import java.time.LocalDate;

import refactown.cleancode.c06objects.book.Geometry;
import refactown.cleancode.c06objects.book.Shape;

public class FeatureEnvy {

	public boolean isValidOrder(AnemicOrder order) {
		return !order.isCancelled() && LocalDate.now().isBefore(order.getDate().plusDays(365)); 
	}
	
	public static double calculateArea(Object shape) {
		if (shape instanceof HibridSquare) {
			HibridSquare s = (HibridSquare)shape;
			return s.getSide() * s.getSide(); // s.side
		} else if (shape instanceof HybridRectangle) {
			HybridRectangle r = (HybridRectangle) shape;
			return r.getHeight() * r.getWidth(); 
		} else if (shape instanceof HybidCircle) {
			HybidCircle c = (HybidCircle) shape;
			return Geometry.PI * c.getRadius() * c.getRadius();
		}
		throw new IllegalArgumentException("Unknown shape class:" + shape.getClass());
	}
}

class HibridSquare implements Shape{

	private double side; // mutable

	public HibridSquare(double side) {
		super();
		this.side = side;
	}

	@Override
	public double area() {
		return side * side;
	}

	public double getSide() {
		return side;
	}

	public void setSide(double side) {
		this.side = side;
	}
}

class HybridRectangle implements Shape {

	private final double height; // immutable
	private final double width; // immutable

	public HybridRectangle(double height, double width) {
		super();
		this.height = height;
		this.width = width;
	}
	
	@Override
	public double area() {
		return height * width;
	}

	public double getHeight() {
		return height;
	}

	public double getWidth() {
		return width;
	}

}

class HybidCircle implements Shape{
	
	private final double radius;

	public HybidCircle(double radius) {
		super();
		this.radius = radius;
	}

	@Override
	public double area() {
		return Geometry.PI * radius * radius;
	}

	public double getRadius() {
		return radius;
	}
}

File: src/main/kotlin/refactown/cleancode/c06objects/FeatureEnvy.kt
package refactown.cleancode.c06objects

import java.lang.IllegalArgumentException
import java.time.LocalDate

class HybridSquare(val side: Double) : Shape {
    override fun area() = side * side
}

class HybridRectangle(val height: Double, val width: Double) : Shape {
    override fun area() = height * width
}

class HybridCircle( val radius: Double) : Shape {
    override fun area() = PI * radius * radius
}

fun calculateAreaFeatureEnvy(shape: Any): Double {
    return when (shape) {
        is HybridSquare -> shape.side * shape.side
        is HybridRectangle -> shape.height * shape.width
        is HybridCircle -> PI * shape.radius * shape.radius
        else -> throw IllegalArgumentException()
    }
}

class AnemicOrder(val date: LocalDate, var cancelled: Boolean)

fun isValidOrder(order:AnemicOrder): Boolean{
    return !order.cancelled && order.date <= order.date.plusDays(365)
}

class Order(val date: LocalDate, var cancelled: Boolean){
    fun isValid()= !cancelled && date <= date.plusDays(365)
}